utils/zerovec/derive/src/ule.rs

```diff
@@ -77,6 +81,254 @@ pub fn derive_impl(input: &DeriveInput) -> TokenStream2 {
+            type GetType = #ule_name;
+            type OwnedType = #name;
+        }
+    )
+}
+
+fn make_ule_enum_impl(
+    name: &Ident,
+    ule_name: &Ident,
+    input: &DeriveInput,
+    enu: &DataEnum,
+) -> TokenStream2 {
+    // We could support more int reprs in the future if needed
+    if !utils::has_valid_repr(&input.attrs, |r| r == "u8") {
+        return Error::new(
+            input.span(),
+            "#[make_ule] can only be applied to #[repr(u8)] enums",
+        )
+        .to_compile_error();
+    }
+
+    for (i, variant) in enu.variants.iter().enumerate() {
+        if variant.fields != Fields::Unit {
+            // This can be supported in the future, see zerovec/design_doc.md
+            return Error::new(
+                variant.span(),
+                "#[make_ule] can only be applied to enums with dataless variants",
+            )
+            .to_compile_error();
+        }
+
+        if let Some((_, ref discr)) = variant.discriminant {
+            if let Some(n) = get_expr_int(discr) {
+                // We require explicit discriminants so that it is clear that reordering
+                // fields would be a breaking change. Furthermore, using explicit discriminants helps ensure that
+                // platform-specific C ABI choices do not matter.
+                // We could potentially add in explicit discriminants on the user's behalf in the future, or support
+                // more complicated sets of explicit discriminant values.
+                if n != i as u64 {
+                    return Error::new(discr.span(), &format!("#[make_ule] must be applied to enums with discriminants \
+                                                              increasing in order from 0, expected {i}, found {n}"))
+                        .to_compile_error();
+                }
+            } else {
+                return Error::new(
+                    discr.span(),
+                    "#[make_ule] must be applied to enums with explicit integer discriminants",
+                )
+                .to_compile_error();
+            }
+        } else {
+            return Error::new(
+                variant.span(),
+                "#[make_ule] must be applied to enums with explicit discriminants",
+            )
+            .to_compile_error();
+        }
+    }
+
+    let max = enu.variants.len() as u8;
+
+    // Safety (based on the safety checklist on the ULE trait):
+    //  1. ULE type does not include any uninitialized or padding bytes.
+    //     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant
+    //  2. ULE type is aligned to 1 byte.
+    //     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant)
+    //  3. The impl of validate_byte_slice() returns an error if any byte is not valid.
+    //  4. The impl of validate_byte_slice() returns an error if there are extra bytes (this never happens)
```

Comment:
```suggestion
    //  3. The impl of validate_byte_slice() returns an error if any byte is not valid.
    //     (Guarantees that the byte is in range of the corresponding enum.)
    //  4. The impl of validate_byte_slice() returns an error if there are extra bytes.
    //     (This does not happen since we are backed by 1 byte.)
```